server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServerResourceManager.java

```diff
@@ -783,6 +788,11 @@ public void executeReadAhead(KeyExtent tablet, ScanDispatcher dispatcher, ScanSe
       scanExecutors.get("meta").execute(task);
     } else {
       DispatchParameters params = new DispatchParamsImpl() {
+
+        // in scan critical path so only create ServiceEnv if needed
+        private final Supplier<ServiceEnvironment> senvSupplier =
+            Suppliers.memoize(() -> new ServiceEnvironmentImpl(context));
```

Comment:
For my own understanding, under which circumstances would this not actually be used, and therefore not need to be created?

Reply:
The [default scan dispatcher](https://github.com/apache/accumulo/blob/5b1c0b54e33ee108a45b2126bd194472703822f2/core/src/main/java/org/apache/accumulo/core/spi/scan/SimpleScanDispatcher.java#L141) does not use the service env for `init()` or `dispatch()`.

Reply:
The `dispatch()` method on the default scan dispatcher is designed to do as little work as possible since its called for each scan batch. Its possible we could remove the memoize for init, because the code only creates new ScanDispatcher objects when table config changes.

Reply:
> Is that right?

That is correct.

> Just a quirk with the current design of the service environment containing snapshots and then caching them.

If it does not exists, probably need better documentation of the lifecycle of an instantiated dispatcher.